File: 3.x.x/crates-utopia/emails/src/canonicals/yandex.rs
use crate::canonicals::provider::{to_lower_case, Canonical, Provider};
use crate::EmailError;

/// Yandex canonicalization (PHP `Yandex`) - case + domain only.
///
/// Not registered in [`crate::Email`]'s provider list (PHP `initializeProviders`
/// omits Yandex); the type is still public and unit-tested like PHP.
#[derive(Debug, Clone, Copy, Default)]
pub struct Yandex;

const SUPPORTED_DOMAINS: &[&str] = &[
    "yandex.ru",
    "yandex.ua",
    "yandex.kz",
    "yandex.com",
    "yandex.by",
    "ya.ru",
];
const CANONICAL_DOMAIN: &str = "yandex.ru";

impl Provider for Yandex {
    fn supports(&self, domain: &str) -> bool {
        SUPPORTED_DOMAINS.contains(&domain)
    }

    fn get_canonical(&self, local: &str, _domain: &str) -> Result<Canonical, EmailError> {
        Ok(Canonical {
            local: to_lower_case(local),
            domain: CANONICAL_DOMAIN.to_string(),
        })
    }

    fn get_canonical_domain(&self) -> &'static str {
        CANONICAL_DOMAIN
    }

    fn get_supported_domains(&self) -> &'static [&'static str] {
        SUPPORTED_DOMAINS
    }
}

File: packages/emails/src/Emails/Canonicals/Providers/Yandex.php
<?php

namespace Utopia\Emails\Canonicals\Providers;

use Utopia\Emails\Canonicals\Provider;

/**
 * Yandex
 *
 * Handles Yandex email normalization based on validator.js rules
 * - Preserves all characters in local part (no subaddress removal)
 * - Normalizes to yandex.ru domain
 */
class Yandex extends Provider
{
    private const SUPPORTED_DOMAINS = [
        'yandex.ru', 'yandex.ua', 'yandex.kz', 'yandex.com', 'yandex.by', 'ya.ru',
    ];

    private const CANONICAL_DOMAIN = 'yandex.ru';

    public function supports(string $domain): bool
    {
        return in_array($domain, self::SUPPORTED_DOMAINS, true);
    }

    public function getCanonical(string $local, string $domain): array
    {
        // Convert to lowercase
        $normalizedLocal = $this->toLowerCase($local);

        // Yandex doesn't remove subaddresses or dots
        // Just normalize case and domain

        return [
            'local' => $normalizedLocal,
            'domain' => self::CANONICAL_DOMAIN,
        ];
    }

    public function getCanonicalDomain(): string
    {
        return self::CANONICAL_DOMAIN;
    }

    public function getSupportedDomains(): array
    {
        return self::SUPPORTED_DOMAINS;
    }
}
